### DISCUSSION TITLE
How to test custom server's routes?

### DISCUSSION THREAD

**Discussion by getaclue00** [vercel/next.js#16476] [Help]
Hello,

I had my server.js like this
```js
const port = parseInt(process.env.PORT, 10) || 3000;
const dev = NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(()=>{
  const server = new Koa();
  const router = new Router();
  //..
  router.all('/(.*)', async (ctx, next) => {
    await handle(ctx.req, ctx.res);
    ctx.respond = false;
    ctx.res.statusCode = 200;
  });

  server.use(router.routes());
  server.use(router.allowedMethods());

  server.listen(port, () => {
    console.log(`> Ready on http://localhost:${port}`);
  });
})
```

I rearranged this code to make it look like so

```js
const port = parseInt(process.env.PORT, 10) || 3000;
const dev = NODE_ENV !== 'production';
const app = next({ dev });
const handle = app.getRequestHandler();

const server = new Koa();
const router = new Router();

app.prepare().then(()=>{
  //..
    const handleStatus = async (ctx, next) => {
      ctx.res.body = JSON.stringify({
        status: 'OK',
      });
      ctx.res.statusCode = 200;
    };

    router.get('/status', handleStatus);

  router.all('/(.*)', async (ctx, next) => {
    await handle(ctx.req, ctx.res);
    ctx.respond = false;
    ctx.res.statusCode = 200;
  });

  server.use(router.routes());
  server.use(router.allowedMethods());
})

module.exports = {server}
```
 
and run my node startup script from a different file like so
```js 
  // startup.js
const { server } = require('./server');
const PORT = process.env.PORT;
const port = parseInt(PORT, 10) || 3000;
const startDB = require('./data/startDB');

async function run() {
  try {
    await startDB();
    server.listen(port, () => {
      console.log(`> Ready on http://localhost:${port}`);
    });
  } catch (err) {
    throw err;
  }
}

run();
```

I thought that this should enable me to run jest and test my routes. But NOPE! 😭 
Cannot seem to fix it. Can someone help me debug please? What did I do wrong?

So my jest looks like this :
```js
// server.test.js
const { server } = require('./server');
const request = require('supertest');
const startDB = require('./data/startDB');

beforeAll(async () => {
  try {
    await startDB();
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
});

it(`gets status endpoint`, async (done) => {
  const response = await request(server.callback()).get('/status');

  expect(response.status).toBe(200);
  done();
});
```

Then I just run `npm run test` and I get this error 🧱

```shell
    Expected: 200
    Received: 404    
    > 24 |   expect(response.status).toBe(200);
         |                           ^
      25 |   done();
      26 | });
```

Any ideas? 
Thanks in advance for any and all help 🤙

**Comment by mustafaKamal-fe** [comment]
155d no answer !!!!! wish there is one 
however, nextjs internals dictate to init custom server as a callback inside `app.prepare().then(...start server)` and this will not work supertest 
will not get the actuall app running the routes. hence there will be no port to perform the test on  (!! perhaps)

**Reply by getaclue00** [reply]
@mustafaKamal-fe if I had an answer I would have updated this :) open to suggestions!

**Reply by mustafaKamal-fe** [reply]
i came up with a temp workaround, 

1. Put your route handlers in a lib folder like usual
2. in another folder create a complete  server, Not nextjs custom server (like expressjs for example) and use route handlers from the lib file in step one
3. finally, put your actuall  nextjs custom server like you would normally for production.

This way i was able to test my route handlers using 'expressj' and 'supertest' (with Jest ) before teh are shipped or used in my prod custom server version.

**Reply by chan360** [reply]
Checking to see if anyone has found a solution to this?

**Comment by imshubhamsingh** [comment]
I solved it using `done` parameter provided in `beforeAll`. You can use this `done` function to ensure that the test only runs when the server is ready. For example, you can put a setTimeout function just after the async function call, and set the interval such that it ensures that the server bootstrap processes are complete.

```js
let server;
beforeAll(async (done) => {
  try {
    await startDB();
    {server} = require('../server')
     // This ensure that Custom Next.js app is started
      setTimeout(() => {
        done();
      }, 80000); // this interval is kept large on purpose.
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
});
```
This is just one of solving this. Another way can be to pass on `done` parameter to an HOF wrapping the server and call the done inside 

```js
function serverHOF(done){
....
app.prepare().then(()=>{ 
    done()
    ....
})
...
}
```
I hope this solves your problem in a much cleaner and efficient way without refactoring much of your code.
